### PR TITLE
SONAR-10147 Remove 'sonar.qualitygate' setting usage to store default quality gate

### PR DESCRIPTION
**Pull request by ehartmann**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SetAsDefaultAction.java [hunk 1]
```diff
@@ -22,14 +22,36 @@
...
+import static org.sonar.server.qualitygate.ws.QualityGatesWsParameters.PARAM_ID;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+import static org.sonarqube.ws.Qualitygates.QualityGate.newBuilder;
 
 public class SetAsDefaultAction implements QualityGatesWsAction {
 
-  private final QualityGates qualityGates;
+  private final DbClient dbClient;
+  private final UserSession userSession;
+  private final QualityGateUpdater qualityGateUpdater;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SetAsDefaultAction.java]: MAJOR Remove this unused "qualityGateUpdater" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java [hunk 2]
```diff
@@ -46,16 +47,16 @@ public QualityGateFinder(DbClient dbClient) {
    * It will first try to get the quality gate explicitly defined on a project, if none it will try to return default quality gate.
    * As it's possible to have no default quality gate, this method can return {@link Optional#empty()}
    */
-  public Optional<QualityGateData> getQualityGate(DbSession dbSession, long componentId) {
+  public Optional<QualityGateData> getQualityGate(DbSession dbSession, OrganizationDto organization, long componentId) {
     Optional<Long> qualityGateId = dbClient.projectQgateAssociationDao().selectQGateIdByComponentId(dbSession, componentId);
     if (qualityGateId.isPresent()) {
       return Optional.of(new QualityGateData(getById(dbSession, qualityGateId.get()), false));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java]: MINOR Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [QualityGateFinder.java#L88](https://github.com/SonarSource/sonarqube/blob/b8c1e1da4863e56330fd69c022fdbbd14a8a1c45/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L88): Remove this unused private "getDefaultId" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MAJOR][MAJOR] [RegisterQualityGates.java#L71](https://github.com/SonarSource/sonarqube/blob/b8c1e1da4863e56330fd69c022fdbbd14a8a1c45/server/sonar-server/src/main/java/org/sonar/server/qualitygate/RegisterQualityGates.java#L71): Remove this unused "qualityGateFinder" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [RegisterQualityGates.java#L72](https://github.com/SonarSource/sonarqube/blob/b8c1e1da4863e56330fd69c022fdbbd14a8a1c45/server/sonar-server/src/main/java/org/sonar/server/qualitygate/RegisterQualityGates.java#L72): Remove this unused "qualityGateUpdater" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)
1. ![MINOR][MINOR] [ShowAction.java#L108](https://github.com/SonarSource/sonarqube/blob/b8c1e1da4863e56330fd69c022fdbbd14a8a1c45/server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java#L108): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QualityGateFinder.java#L67](https://github.com/SonarSource/sonarqube/blob/b8c1e1da4863e56330fd69c022fdbbd14a8a1c45/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L67): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L23](https://github.com/SonarSource/sonarqube/blob/c80d3f4c388a0d557df9ce39282bf2ad15611624/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L23): Remove this unused import 'org.apache.commons.lang.StringUtils'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [QualityGateFinder.java#L27](https://github.com/SonarSource/sonarqube/blob/c80d3f4c388a0d557df9ce39282bf2ad15611624/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L27): Remove this unused import 'org.sonar.db.property.PropertyDto'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [ShowAction.java#L108](https://github.com/SonarSource/sonarqube/blob/c80d3f4c388a0d557df9ce39282bf2ad15611624/server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java#L108): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QualityGateFinder.java#L67](https://github.com/SonarSource/sonarqube/blob/c80d3f4c388a0d557df9ce39282bf2ad15611624/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L67): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L23](https://github.com/SonarSource/sonarqube/blob/cdb463ffd7d050336b43363f2e09058888746ca1/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L23): Remove this unused import 'org.apache.commons.lang.StringUtils'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [QualityGateFinder.java#L27](https://github.com/SonarSource/sonarqube/blob/cdb463ffd7d050336b43363f2e09058888746ca1/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L27): Remove this unused import 'org.sonar.db.property.PropertyDto'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [ShowAction.java#L108](https://github.com/SonarSource/sonarqube/blob/cdb463ffd7d050336b43363f2e09058888746ca1/server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java#L108): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QualityGateFinder.java#L67](https://github.com/SonarSource/sonarqube/blob/cdb463ffd7d050336b43363f2e09058888746ca1/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L67): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ShowAction.java#L108](https://github.com/SonarSource/sonarqube/blob/c40ce0aa6c92ead8b13723af1fc14861c11de5f3/server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java#L108): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QualityGateFinder.java#L65](https://github.com/SonarSource/sonarqube/blob/c40ce0aa6c92ead8b13723af1fc14861c11de5f3/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L65): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ShowAction.java#L108](https://github.com/SonarSource/sonarqube/blob/4cd7125e4cca712efc2f30794607447416b2ee5a/server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java#L108): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![INFO][INFO] [QualityGateFinder.java#L65](https://github.com/SonarSource/sonarqube/blob/4cd7125e4cca712efc2f30794607447416b2ee5a/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L65): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateUpdater.java [hunk 3]
```diff
@@ -73,15 +68,9 @@ private void validateQualityGate(DbSession dbSession, OrganizationDto organizati
...
-    if (qualityGateDto == null) {
-      dbClient.propertiesDao().deleteGlobalProperty(SONAR_QUALITYGATE_PROPERTY, dbSession);
-    } else {
-      long qualityGateId = qualityGateDto.getId();
-      checkFound(dbClient.qualityGateDao().selectById(dbSession, qualityGateId), "There is no quality gate with id=" + qualityGateId);
-      dbClient.propertiesDao().saveProperty(dbSession,
-        new PropertyDto().setKey(SONAR_QUALITYGATE_PROPERTY).setValue(qualityGateDto.getId().toString()));
-    }
+  public void setDefault(DbSession dbSession, OrganizationDto organizationDto, QualityGateDto qualityGateDto) {
+    organizationDto.setDefaultQualityGateUuid(qualityGateDto.getUuid());
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateUpdater.java]: It should be best to have only one way to set the default quality gate on an organization : either we use OrganizationDao#setDefaultQualityGate, or OrganizationDao#update, but not both

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SetAsDefaultAction.java [hunk 4]
```diff
@@ -45,11 +67,23 @@ public void define(WebService.NewController controller) {
...
   @Override
   public void handle(Request request, Response response) {
-    qualityGates.setDefault(QualityGatesWs.parseId(request, QualityGatesWsParameters.PARAM_ID));
+    Long id = parseId(request, PARAM_ID);
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      OrganizationDto organization = wsSupport.getOrganization(dbSession, request);
+      userSession.checkPermission(ADMINISTER_QUALITY_GATES, organization);
+      QualityGateDto qualityGate = qualityGateFinder.getByOrganizationAndId(dbSession, organization, id);
+      organization.setDefaultQualityGateUuid(qualityGate.getUuid());
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SetAsDefaultAction.java]: I think we should use OrganizationDao#setDefaultQualityGate

File: server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java [hunk 5]
```diff
@@ -81,15 +81,15 @@ private static ComponentMapper mapper(DbSession session) {
...
+    java.util.Optional<ComponentDto> componentDto = selectById(session, id);
     if (!componentDto.isPresent()) {
       throw new RowNotFoundException(String.format("Component id does not exist: %d", id));
     }
     return componentDto.get();
   }
 
-  public Optional<ComponentDto> selectById(DbSession session, long id) {
-    return Optional.fromNullable(mapper(session).selectById(id));
+  public java.util.Optional<ComponentDto> selectById(DbSession session, long id) {
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java]: Be careful, this method is probably used by core plugins

File: server/sonar-db-dao/src/main/resources/org/sonar/db/organization/OrganizationMapper.xml [hunk 6]
```diff
@@ -7,6 +7,7 @@
     org.kee as "key",
     org.name as "name",
     org.description as "description",
+    org.default_quality_gate_uuid as "defaultQualityGateUuid",
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/organization/OrganizationMapper.xml]: This should not be needed, using QualityGateDao#selectDefault should be enough to get the default quality gate of an organization

File: server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java (hunk 5 above)
Comment (julienlancelot) [reply server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDao.java]: Moreover, it would have been best to have it in a separate commit :)

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java [hunk 7]
```diff
@@ -46,16 +47,16 @@ public QualityGateFinder(DbClient dbClient) {
...
     } else {
-      Optional<QualityGateDto> defaultQualityGate = getDefault(dbSession);
-      if (!defaultQualityGate.isPresent()) {
+      QualityGateDto defaultQualityGate = dbClient.qualityGateDao().selectByOrganizationAndUuid(dbSession, organization, organization.getDefaultQualityGateUuid());
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java]: Please use qualityGateDao#getDefault

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java [hunk 8]
```diff
@@ -46,16 +47,16 @@ public QualityGateFinder(DbClient dbClient) {
...
+      if (defaultQualityGate == null) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java]: We should fail it this case

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java [hunk 9]
```diff
@@ -72,16 +73,10 @@ public QGateWithOrgDto getByOrganizationAndId(DbSession dbSession, OrganizationD
       "No quality gate has been found for id %s in organization %s", qualityGateId, organization.getName());
   }
 
-  public Optional<QualityGateDto> getDefault(DbSession dbSession) {
-    Optional<Long> defaultQualityGateId = getDefaultId(dbSession);
+  public Optional<QualityGateDto> getDefault(DbSession dbSession, OrganizationDto organization) {
+    QualityGateDto defaultQualityGate = dbClient.qualityGateDao().selectByOrganizationAndUuid(dbSession, organization, organization.getDefaultQualityGateUuid());
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java]: Please use qualityGateDao#getDefault

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateUpdater.java [hunk 10]
```diff
@@ -73,6 +76,17 @@ public void setDefault(DbSession dbSession, OrganizationDto organizationDto, Qua
     dbClient.qualityGateDao().update(qualityGateDto, dbSession);
   }
 
+  public void dissociateProject(DbSession dbSession, ComponentDto project) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateUpdater.java]: This should have been done in another commit

File: server/sonar-db-dao/src/test/java/org/sonar/db/qualitygate/QualityGateDbTester.java [hunk 11]
```diff
@@ -86,17 +77,21 @@ public void associateProjectToQualityGate(ComponentDto component, QualityGateDto
     db.commit();
   }
 
-  public QualityGateDto createDefaultQualityGate(String qualityGateName) {
-    QualityGateDto defaultQGate = insertQualityGate(qualityGateName);
-    setDefaultQualityGate(defaultQGate);
+  public void associateQualityGateToOrganization(QualityGateDto qualityGate, OrganizationDto organization) {
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/test/java/org/sonar/db/qualitygate/QualityGateDbTester.java]: Why do we need this ?

Comment (julienlancelot) [reply server/sonar-db-dao/src/test/java/org/sonar/db/qualitygate/QualityGateDbTester.java]: As discussed, this method is not used and should be removed

File: server/sonar-db-dao/src/test/java/org/sonar/db/qualitygate/QualityGateDbTester.java [hunk 12]
```diff
@@ -86,17 +77,21 @@ public void associateProjectToQualityGate(ComponentDto component, QualityGateDto
...
+    dbClient.qualityGateDao().associate(dbSession, Uuids.createFast(), organization, qualityGate);
+    dbSession.commit();
+  }
+
+  @SafeVarargs
+  public final QualityGateDto createDefaultQualityGate(OrganizationDto organization, Consumer<QualityGateDto>... dtoPopulators) {
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/test/java/org/sonar/db/qualitygate/QualityGateDbTester.java]: I think using insert() then setDefault() should be enough

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java [hunk 13]
```diff
@@ -72,16 +73,10 @@ public QGateWithOrgDto getByOrganizationAndId(DbSession dbSession, OrganizationD
       "No quality gate has been found for id %s in organization %s", qualityGateId, organization.getName());
   }
 
-  public Optional<QualityGateDto> getDefault(DbSession dbSession) {
-    Optional<Long> defaultQualityGateId = getDefaultId(dbSession);
+  public Optional<QualityGateDto> getDefault(DbSession dbSession, OrganizationDto organization) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java]: We should not return an Optional, as it's not possible to have no default QG

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/GetByProjectAction.java [hunk 14]
```diff
@@ -90,8 +93,13 @@ public void handle(Request request, Response response) throws Exception {
         !userSession.hasComponentPermission(UserRole.ADMIN, project)) {
         throw insufficientPrivilegesException();
       }
+      Optional<OrganizationDto> organization = dbClient.organizationDao().selectByUuid(dbSession, project.getOrganizationUuid());
 
-      Optional<QualityGateData> data = qualityGateFinder.getQualityGate(dbSession, project.getId());
+      if (!organization.isPresent()) {
+        throw new NotFoundException(format("Organization [%s] is absent", project.getOrganizationUuid()));
+      }
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/GetByProjectAction.java]: Please use QualityGateWsSupport#getOrganization (QualityGateWsSupport is already used in this class, you only need to rebase on the common branch)

**Comment by julienlancelot** [comment]
The drop of the "sonar.qualitygate" setting at global level is not done in DB

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java [hunk 15]
```diff
@@ -44,19 +44,16 @@ public QualityGateFinder(DbClient dbClient) {
...
-      return Optional.of(new QualityGateData(qualityGate, false));
+      return new QualityGateData(qualityGate, false);
     } else {
-      Optional<QualityGateDto> defaultQualityGate = getDefault(dbSession);
-      if (!defaultQualityGate.isPresent()) {
-        return Optional.empty();
-      }
-      return Optional.of(new QualityGateData(defaultQualityGate.get(), true));
+      QualityGateDto defaultQualityGate = dbClient.qualityGateDao().selectByOrganizationAndUuid(dbSession, organization, organization.getDefaultQualityGateUuid());
+      checkState(defaultQualityGate != null, format("Unable to find the quality gate [%s] for organization [%s]", organization.getDefaultQualityGateUuid(), organization.getUuid()));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java]: MAJOR Invoke method(s) only conditionally. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)
MINOR Split this 181 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'
